multi/gr716-multi/gpio.c

```diff
@@ -0,0 +1,226 @@
+
+#define GPIO_PORT_CNT 2
+
+#define GRGPIO0_BASE ((void *)0x8030C000)
+#define GRGPIO1_BASE ((void *)0x8030D000)
+
+/* GPIO registers */
+
+#define GPIO_DATA      0       /* Port data reg : 0x00 */
+#define GPIO_OUT       1       /* Output reg : 0x04 */
+#define GPIO_DIR       2       /* Port direction reg : 0x08 */
+#define GPIO_IMASK     3       /* Interrupt mask reg : 0x0C */
+#define GPIO_IPOL      4       /* Interrupt polarity reg : 0x10 */
+#define GPIO_IEDGE     5       /* Interrupt edge reg : 0x14 */
+#define GPIO_CAP       7       /* Port capability reg : 0x1C */
+#define GPIO_IRQMAP(n) (8 + n) /* Interrupt map register n : 0x20 - 0x3C */
+#define GPIO_IAVAIL    16      /* Interrupt available reg : 0x40 */
+#define GPIO_IFLAG     17      /* Interrupt flag reg : 0x44 */
+#define GPIO_IPEN      18      /* Interrupt enable reg : 0x48 */
+#define GPIO_PULSE     19      /* Pulse reg : 0x4C */
+#define GPIO_IE_LOR    20      /* Interrupt enable logical OR reg : 0x50 */
+#define GPIO_PO_LOR    21      /* Port output logical OR reg : 0x54 */
+#define GPIO_PD_LOR    22      /* Port direction logical OR reg : 0x58 */
+#define GPIO_IM_LOR    23      /* Interrupt mask logical OR reg : 0x5C */
+#define GPIO_IE_LAND   24      /* Interrupt enable logical AND reg : 0x60 */
+#define GPIO_PO_LAND   25      /* Port output logical AND reg : 0x64 */
+#define GPIO_PD_LAND   26      /* Port direction logical AND reg : 0x68 */
+#define GPIO_IM_LAND   27      /* Interrupt mask logical AND reg : 0x6C */
+#define GPIO_IE_LXOR   28      /* Interrupt enable logical XOR reg : 0x70 */
+#define GPIO_PO_LXOR   29      /* Port output logical XOR reg : 0x74 */
+#define GPIO_PD_LXOR   30      /* Port direction logical XOR reg : 0x78 */
+#define GPIO_IM_LXOR   31      /* Interrupt mask logical XOR reg : 0x7C */
+#define GPIO_IE_SC     32      /* Interrupt enable logical set/clear reg : 0x80 - 0x8C */
+#define GPIO_PO_SC     36      /* Port output logical set/clear reg : 0x90 - 0x9C */
+#define GPIO_PD_SC     40      /* Port direction logical set/clear reg : 0xA0 - 0xAC */
+#define GPIO_IM_SC     44      /* Interrupt mask logical set/clear reg : 0xB0 - 0xBC */
+
+
+static struct {
+	volatile uint32_t *grgpio0;
+	volatile uint32_t *grgpio1;
+	handle_t lock;
+} gpio_common;
+
+
+static inline int gpio_pinToPort(uint8_t pin)
+{
+	return (pin >> 5);
+}
+
+
+static inline volatile uint32_t *gpio_portToAddr(int port)
+{
+	switch (port) {
+		case GPIO_PORT_0:
+			return gpio_common.grgpio0;
+		case GPIO_PORT_1:
+			return gpio_common.grgpio1;
+		default:
+			return NULL;
+	}
+}
+
+
+static int gpio_setPort(int port, uint32_t mask, uint32_t val)
+{
+	uint32_t clearedMsk;
+	uint32_t set = val & mask;
+	uint32_t clear = (~val) & mask;
+
+	volatile uint32_t *gpio = gpio_portToAddr(port);
+
+	if (gpio == NULL) {
+		return -EINVAL;
+	}
+
+	mutexLock(gpio_common.lock);
+	clearedMsk = *(gpio + GPIO_OUT) & ~clear;
+	*(gpio + GPIO_DIR) = clearedMsk | set;
+	mutexUnlock(gpio_common.lock);
```

Comment:
I guess it's not possible to perform atomic `and` and `or` here (from `common`)? This would allow us to skip mutex

Reply:
If it would be possible in case of direction too, mutex could be removed altogether

Reply:
well, I tried to use it but I think it didn't want to work, I'll check again

Reply:
I couldn't get atomic set & clear to work, but I use gpio logical `or` and `and` registers now